### DISCUSSION TITLE
Why my code doesn't work, just getter/setter, cause ui cannot listen the state change

### DISCUSSION THREAD

**Discussion by liuhuapiaoyuan** [pmndrs/valtio#862] [Bug report]
## Summary
Why my code doesn't work, just getter/setter, cause ui cannot listen the state change

```typescript
import { Slider } from '@nextui-org/react'
import React, { useContext, useEffect, useRef, useState } from 'react'
import { twMerge } from 'tailwind-merge'
import { proxy, ref, useSnapshot } from 'valtio'
import { Drawer } from './lib/drawer'
import { WhiteboardStateContext } from './lib/model'

export type DrawContentProps = {
  className?: string
  style?: React.CSSProperties
}
export class WhiteboardState {
  /**
   * 尝试代理drawer
   */
  draw?: Drawer
  counter: number = 0

  constructor(data?: Partial<WhiteboardState>) {
    this.brushWidth = data?.brushWidth ?? 5
    this.brushColor = data?.brushColor ?? 'black'
    this.selectedShape = data?.selectedShape ?? 'null'
    this.draw = data?.draw
  }

  get brushWidth() {
    return this.draw?.brushWidth ?? 0
  }
  set brushWidth(width: number) {
    console.log('WhiteboardState', 'brushWidth', width, this.counter)
    if (this.draw) {
      this.draw.brushWidth = width
    }
  }
  get brushColor() {
    return this.draw?.brushColor ?? 'black'
  }
  set brushColor(brushColor: string) {
    if (this.draw) {
      this.draw.brushColor = brushColor
    }
  }
  get selectedShape() {
    return this.draw?.selectedShape ?? ''
  }
  set selectedShape(selectedShape: string) {
    if (this.draw) {
      this.draw.selectedShape = selectedShape
    }
  }
}

function PenSlider() {
  const model = useContext(WhiteboardStateContext)
  const state = useSnapshot(model)
  return (
    <Slider
      value={state.brushWidth}
      onChange={e => {
        model.brushWidth = e as number
      }}
    />
  )
}

function DrawContent(props: { draw: Drawer }) {
  const { draw } = props
  const state = useRef(proxy(new WhiteboardState({ draw: ref(draw) }))).current
  return (
    <WhiteboardStateContext.Provider value={state}>
      <>
        <div className='absolute left-0 top-0 h-full flex items-center'>
          <PenSlider />
        </div>
      </>
    </WhiteboardStateContext.Provider>
  )
}

export default function Whiteboard(props: DrawContentProps) {
  const { className, style } = props
  const canvansRef = React.useRef<HTMLCanvasElement>(null)
  const containerRef = React.useRef<HTMLDivElement>(null)
  const [draw, setDraw] = useState<Drawer>()

  useEffect(() => {
    if (!canvansRef.current || !containerRef.current) return
    const canvas = new Drawer(canvansRef.current, containerRef.current)
    setDraw(canvas)
    return () => {
      canvas.destroy()
      setDraw(undefined)
    }
  }, [canvansRef])

  return (
    <div
      ref={containerRef}
      className={twMerge('w-full relative h-full', className)}
      style={style}
    >
      <canvas ref={canvansRef} />
      {draw && <DrawContent draw={draw} />}
    </div>
  )
}

```

**Comment by dai-shi** [comment]
Can you create a **minimal** reproduction with stackblitz or something?

**Reply by liuhuapiaoyuan** [reply]
Of course, it's possible
https://stackblitz.com/edit/stackblitz-starters-5kjbzt?file=app%2Fcomponents%2FPane.tsx

**Reply by liuhuapiaoyuan** [reply]
Please help me research, thank you very much

**Reply by dai-shi** [reply]
You can make it smaller without styles and some others. Remove all code that is not related to the reproduction. And, if possible, just use one file please. As few lines as possible.
At a glance, using `ref()` may prevent the rerender.

**Reply by liuhuapiaoyuan** [reply]
I think this is pretty minimal already.. BTW, if we don't use ref, how do we connect with an external class instance?

**Reply by liuhuapiaoyuan** [reply]
You are right. If ref is removed I can respond, but how can I associate to a very complex class instance?
If I do not use ref, then valitio gives an error

**Reply by liuhuapiaoyuan** [reply]
I have also created a smaller single file:

https://stackblitz.com/edit/stackblitz-starters-du88wa?file=src%2FApp.tsx

**Reply by dai-shi** [reply]
> I think this is pretty minimal already..

Maybe it was from your perspective. But it's not from the other perspective.

https://stackblitz.com/edit/stackblitz-starters-du88wa?file=src%2FApp.tsx is better.

> If ref is removed I can respond

Yeah, ref is to avoid responding.

> If I do not use ref, then valitio gives an error

If ref is required, I think you need subscription mechanism on your end.
Like, if `_width` changes, you change `other`. Check this out:
https://stackblitz.com/edit/stackblitz-starters-lmfugz?file=src%2FApp.tsx

**Reply by liuhuapiaoyuan** [reply]
Thank you very much. But in that case, I changed state, and it doesn't synch to my content, I need a mechanism for bi-bound, 🙏

**Reply by dai-shi** [reply]
yeah, good point.

**Reply by liuhuapiaoyuan** [reply]
> yeah, good point.

Any  ideas?You can imagine a complex 3D Comer, a Figure Comer, that needs to be in the UI, what a common situation.

**Reply by liuhuapiaoyuan** [reply]
I think the ref mechanism is good,I'm also free to use the getter / setter

**Reply by dai-shi** [reply]
I think it's easier to create a method to update `width`.

**Reply by liuhuapiaoyuan** [reply]
Well, thank you very much.
